crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -457,36 +305,161 @@ impl Drop for AbortOnDrop {
+            consumer_group_id,
+            failed,
+        }
+    }
+
+    async fn run(mut self) {
+        // this future will be aborted when the partition is no longer needed, so any exit is a failure
+        if let Err(err) = self.inner().await {
+            _ = self.failed.send(err);
         }
-        if let Err(err) = consumer.store_offset(&topic_partition.0, topic_partition.1, offset) {
-            break err.into();
+    }
+
+    async fn inner(&mut self) -> Result<(), Error> {
+        debug!(
+            restate.subscription.id = %self.builder.subscription().id(),
+            messaging.consumer.group.name = self.consumer_group_id,
+            "Starting topic '{}' partition '{}' consumption loop",
+            self.topic_partition.0,
+            self.topic_partition.1
+        );
+
+        let producer_id = dedup_producer_id(
+            &self.consumer_group_id,
+            &self.topic_partition.0,
+            self.topic_partition.1,
+        );
+
+        let ingress_request_counter = counter!(
+            KAFKA_INGRESS_REQUESTS,
+            "subscription" => self.builder.subscription().id().to_string(),
+            "topic" => self.topic_partition.0.to_string(),
+            "partition" => self.topic_partition.1.to_string(),
+        );
+
+        let mut inflight = VecDeque::new();
+
+        loop {
+            tokio::select! {
+                biased;
+                Some(committed) = Self::head_committed(&mut inflight) => {
+                    let head = inflight.pop_front().expect("to exist");
+                    if let Err(CommitError::Cancelled) = committed {
+                        return Err(Error::IngressClosed);
+                    }
+                    ingress_request_counter.increment(1);
+                    trace!(
+                        topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1, offset=%head.offset,
+                        "store kafka offset",
+                    );
+                    self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, head.offset)?;
+                },
+                received = self.topic_partition_consumer.recv() => {
+                    trace!(
+                        topic=%self.topic_partition.0, kafka_partition=%self.topic_partition.1,
+                        "ingesting kafka message"
+                    );
+
+                    let msg = received?;
+
+                    let offset = msg.offset();
+
+                    let envelope = self.builder.build(producer_id, &self.consumer_group_id, msg)?;
+                    let pk = envelope.partition_key();
+                    let msg = IngestRecord::from_parts(envelope.record_keys(), envelope);
+                    // while trying to send, we also make sure to process
+                    // committed messages to release permits as soon as
+                    // possible, which gives better chances for sending to
+                    // not block
+
```

Comment:
nit: maybe add line break above the comment and remove line break before loop (assuming that the comment belongs to the loop).